pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,368 @@
+			return nil, fmt.Errorf("error retrieving AWS cloud provider")
+		}
+		pvlc.ebsVolumes = awsCloudProvider
+	}
+	return pvlc.ebsVolumes, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findGCEPDLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.gceCloudProvider == nil {
+		gceCloudProvider, ok := pvlc.cloud.(*gce.GCECloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving GCE cloud provider")
+		}
+		pvlc.gceCloudProvider = gceCloudProvider
+	}
+	return pvlc.gceCloudProvider, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) updateVolume(volName string, volLabels map[string]string) error {
+	glog.V(4).Infof("updating PersistentVolume %s", volName)
+	curVol, err := pvlc.kubeClient.Core().PersistentVolumes().Get(volName, metav1.GetOptions{IncludeUninitialized: true})
+	if err != nil {
+		return err
+	}
+	objCopy := curVol.DeepCopyObject()
+	newVolume, ok := objCopy.(*v1.PersistentVolume)
+	if !ok {
+		return fmt.Errorf("failed to cast copy into persistentvolume object %#v", newVolume)
+	}
+	if newVolume.Labels == nil {
+		newVolume.Labels = make(map[string]string)
+	}
+	for k, v := range volLabels {
+		newVolume.Labels[k] = v
+	}
+	pvlc.removeInitializer(newVolume)
+
+	oldData, err := json.Marshal(curVol)
+	if err != nil {
+		return fmt.Errorf("failed to marshal old persistentvolume %#v for persistentvolume %q: %v", curVol, volName, err)
+	}
+
+	newData, err := json.Marshal(newVolume)
+	if err != nil {
+		return fmt.Errorf("failed to marshal new persistentvolume %#v for persistentvolume %q: %v", newVolume, volName, err)
+	}
+
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.PersistentVolume{})
+	if err != nil {
+		return fmt.Errorf("failed to create patch for persistentvolume %q: %v", volName, err)
+	}
+
+	_, err = pvlc.kubeClient.Core().PersistentVolumes().Patch(string(volName), types.StrategicMergePatchType, patchBytes)
+	if err != nil {
+		return fmt.Errorf("failed to update PersistentVolume %s: %v", volName, err)
+	}
+	glog.V(4).Infof("updated PersistentVolume %s", volName)
+
+	return err
+}
+
+func (pvlc *PersistentVolumeLabelController) removeInitializer(pv *v1.PersistentVolume) {
```

Comment:
rather than mutate the arg, how about taking the slice of initializers and returning a new slice of initializers to assign.  That will make for a method that can later be a library method.

Reply:
Done